dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt

```diff
@@ -37,13 +56,219 @@ import okhttp3.OkHttpClient
+            intercept(chain, request, tracer)
+        }
+    }
+
+    // endregion
+
+    // region TracingInterceptor
+
+    /**
+     * Called whenever a span was successfully created around an OkHttp [Request].
+     * The given [Span] can be updated (e.g.: add custom tags / baggage items) before it is
+     * finalized.
+     * @param request the intercepted [Request]
+     * @param span the [Span] created around the [Request]
+     * @param response the [Request] response (or null if an error occurred)
+     * @param throwable the error which occurred during the [Request] (or null)
+     */
+    protected open fun onRequestIntercepted(
+        request: Request,
+        span: Span,
+        response: Response?,
+        throwable: Throwable?
+    ) {
+        tracedRequestListener.onRequestIntercepted(request, span, response, throwable)
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun isRequestTraceable(request: Request): Boolean {
+        val host = request.url().host()
+        return host.matches(hostRegex)
+    }
+
+    @Suppress("TooGenericExceptionCaught", "ThrowingInternalException")
+    private fun intercept(
+        chain: Interceptor.Chain,
+        request: Request,
+        tracer: Tracer
+    ): Response {
+        val span = buildSpan(tracer, request)
+        val updatedRequest = updateRequest(request, tracer, span).build()
+
+        try {
+            val response = chain.proceed(updatedRequest)
+            handleResponse(request, response, span)
+            return response
+        } catch (e: Throwable) {
+            handleThrowable(request, e, span)
+            throw e
+        }
+    }
+
+    @Synchronized
+    private fun resolveTracer(): Tracer? {
+        return if (!TracesFeature.initialized.get()) {
+            devLogger.w(WARNING_TRACING_DISABLED)
+            null
+        } else if (GlobalTracer.isRegistered()) {
+            // clear the localTracer reference if any
+            localTracerReference.set(null)
+            GlobalTracer.get()
+        } else {
+            // we check if we already have a local tracer if not we instantiate one
+            resolveLocalTracer()
+        }
+    }
+
+    private fun resolveLocalTracer(): Tracer {
+        if (localTracerReference.get() == null) {
+            // only register once
+            localTracerReference.compareAndSet(null, localTracerFactory())
+            devLogger.w(WARNING_DEFAULT_TRACER)
+        }
+        return localTracerReference.get()
+    }
+
+    private fun buildSpan(tracer: Tracer, request: Request): Span {
+        val parentContext = extractParentContext(tracer, request)
+        val url = request.url().toString()
+
+        val span = tracer.buildSpan(SPAN_NAME)
+            .asChildOf(parentContext)
+            .start()
+
+        (span as? MutableSpan)?.resourceName = url
+        span.setTag(Tags.HTTP_URL.key, url)
+        span.setTag(Tags.HTTP_METHOD.key, request.method())
+
+        return span
```

Comment:
can you please explain a little bit how we are going to use `RUMResourceID` here?

Reply:
Actually this will be done in the DatadogInterceptor. The nice thing about spans is that you can add a tag to it whenever you want, so in the `onRequestIntercepted` method, we can inject the resourceId as soon as we know that there will be a resource event (and not an error event).
Also this will allow us to discard spans on resources that shouldn't be traced.